### PR TITLE
Paginate list of Circles through OCS query parameters

### LINKED ISSUE #902: Circles list pagination
The OCS route `/ocs/v2.php/apps/circles/circles` returns all circles on the instance.

I have a production instance with nearly a thousand circles, so it's quite an expensive query to perform. I'd like a way to paginate the circles list to avoid such a query and response payload.

### PR DESCRIPTION
**Pull request by tcitworld**
Closes #902

### CODE REVIEW

File: lib/Controller/LocalController.php [hunk 1]
```diff
@@ -411,10 +411,12 @@ public function memberRemove(string $circleId, string $memberId): DataResponse {
 	/**
 	 * @NoAdminRequired
 	 *
+	 * @param int|null $limit
+	 * @param int|null $offset
 	 * @return DataResponse
 	 * @throws OCSException
 	 */
-	public function circles(): DataResponse {
+	public function circles(int $limit = null, int $offset = null): DataResponse {
```
Comment (ArtificialOwl) [inline lib/Controller/LocalController.php:419]: I would go with `int $limit = -1, int $offset = 0`

File: lib/Controller/LocalController.php [hunk 2]
```diff
@@ -423,6 +425,13 @@ public function circles(): DataResponse {
 				  ->filterBackendCircles()
 				  ->addDetail(BasicProbe::DETAILS_POPULATION);
 
+			if ($limit !== null) {
```
Comment (ArtificialOwl) [inline lib/Controller/LocalController.php:428]: and remove conditions

File: lib/Controller/LocalController.php [hunk 3]
```diff
@@ -423,6 +425,13 @@ public function circles(): DataResponse {
...
+				$probe->setItemsLimit($limit);
+			}
+			if ($offset !== null) {
```
Comment (ArtificialOwl) [inline lib/Controller/LocalController.php:431]: (both)

**Comment by ArtificialOwl** [comment]
Can you also add this in the `AdminController.php` ?

**Review by ArtificialOwl — APPROVED** [review]

**Comment by tcitworld** [comment]
I wanted to use this for https://github.com/nextcloud/contacts/issues/2605, which doesn't really do BC with NC major versions, so I guess so, yes.

#919 #920

**Pull request merged**